pdns/dnsdistdist/dnsdist-opentelemetry.hh

```diff
@@ -0,0 +1,318 @@
+   */
+  [[nodiscard]] TraceID getTraceID() const;
+
+  /**
+   * @brief Generate the TracesData from all data in this Tracer
+   *
+   * @return pdns::trace::TracesData
+   */
+  [[nodiscard]] TracesData getTracesData() const;
+
+  /**
+   * @brief Get the TracesData as protobuf encoded OpenTelemetry data
+   */
+  [[nodiscard]] std::string getOTProtobuf() const;
+
+  /**
+   * @class Closer
+   * @brief Automatically closes a Span when it goes out of scope
+   *
+   * This is a helper that _somewhat_ implements Go's `defer` in C++ semantics
+   * Basically, it stores a pointer to the Tracer and a SpanID.
+   * When the object goes our of scope, the closeSpan function is called
+   */
+  class Closer
+  {
+  public:
+    /**
+     * @brief An empty Closer, not really useful
+     */
+#ifdef DISABLE_PROTOBUF
+    Closer() = default;
+#else
+    Closer() :
+      d_tracer(nullptr), d_spanID(SpanID{}) {};
+    /**
+     * @brief Create a Closer
+     *
+     * There should be no need to call this directly. Use one of these functions to get one:
+     *
+     * Tracer::getCloser
+     * Tracer::openSpan
+     *
+     * @param tracer A pointer to the Tracer where we want to close a Span
+     * @param spanid The SpanID to close in the Tracer
+     */
+    Closer(Tracer* tracer, const SpanID& spanid) :
+      d_tracer(tracer), d_spanID(spanid) {};
+#endif
+
+    /**
+     * @brief Closes the Span in the Tracer
+     */
+    ~Closer()
+    {
+#ifndef DISABLE_PROTOBUF
+      if (d_tracer != nullptr) {
+        d_tracer->closeSpan(d_spanID);
+      }
+#endif
+    };
+    Closer(const Closer&) = default;
+    Closer& operator=(const Closer&) = default;
+    Closer& operator=(Closer&&) noexcept = default;
+    Closer(Closer&&) = default;
+
+    /**
+     * @brief Get the SpanID
+     *
+     * @return
+     */
+    [[nodiscard]] SpanID getSpanID() const;
+
+  private:
+#ifndef DISABLE_PROTOBUF
+    // XXX: Should we make this a shared_ptr and force all consumers to Tracer to keep it as a shared_ptr as well?
+    Tracer* d_tracer;
+    SpanID d_spanID;
+#endif
+  };
+
+  /**
+   * @brief Get a Closer for spanid in this Tracer
+   *
+   * @param spanid The SpanID that will close when the Closer is destructed
+   * @return Tracer::Closer
+   */
+  Closer getCloser(const SpanID& spanid);
+
+  /**
+   * @brief Add a new Span
+   *
+   * @param name The name for this span
+   * @return Tracer::Closer for the newly created Span
+   */
+  Closer openSpan(const std::string& name);
+
+  /**
+   * @brief Add a new Span which is a child of another Span
+   *
+   * @param name The name for this span
+   * @param parentSpanID The SpanID of the parent Trace
+   * @return Tracer::Closer for the newly created Span
+   */
+  Closer openSpan(const std::string& name, const SpanID& parentSpanID);
+
+private:
+#ifndef DISABLE_PROTOBUF
+  /**
+   * @brief Whether or not we are storing full Spans or minimal Spans
+   */
+  bool d_activated{false};
```

Comment:
Struct padding is not going to be great, but let's not worry about that until we have taken care of everything else.